### PR TITLE
Reduce the memory consumption of nodes v2

### PR DESCRIPTION
**Pull request by sethtroisi**
Status: **Fixing problems with get_children()**

This is an updated version of @ywrt  https://github.com/gcp/leela-zero/pull/574

The core idea is to only create `UCTNode` children nodes (which cost ~60 bytes of memory) for the root node and just before they are used in `play_simulation` otherwise children (a move and a score) are stored in a `vector<pair<float, short>>` of their parent (~6 bytes).

**Results**
* **66%** reduction in memory (From ~60 Bytes to ~20 Bytes) 
* increase in max_tree_size from **100 to 300 million positions** (allowing for closer to 1M playouts per position, maybe ~500K with tree reuse)
* **+6-12% speed**, measured with --benchmark -p 25000 on three machines

This change adds more complexity to UCTNode (having to lock the mutex in a many more methods, having `expand` calls mixed in with uct_select_child). As UCTNode becomes more complex the risk of low frequency rarely seen threading bugs increases. After this is checked in I promise to do a pass through UCTSearch and UCTNode and see if I can reduce the complexity and number of methods

----
**Testing** all tests performed in both debug and normal build

**--benchmark** matches for several thousand but then diverges because of network eval.
`./leelaz -w weights --noponder --benchmark -p 2323`

Hand inspected the first `uct_select_child` instance that don't match with both gpu and cpu paths, difference is in 5th sig fig of winrate calculation

0.519917 = 0.503896 + 0.234812, 564 D16
0.519903 = 0.503882 + 0.234812, 564 D16

**--benchmark** with -p 10000 and -p 100,000 to verify memory usage

	Command being timed: "./leelaz -w weights --noponder --benchmark -p 10000"
	Maximum resident set size (kbytes): 178920
        vs
	Maximum resident set size (kbytes): 388100
        and
	Maximum resident set size (kbytes): 612792
        vs
	Maximum resident set size (kbytes): 2686648

**low playouts**
`(yes "go" | head -n 60) | ./leelaz -w weights -s1 -t1 --noponder -p {1,2,5,10,20}`
Hash and Ko-Hash match for all playouts
(this discovered https://github.com/gcp/leela-zero/pull/948)

**Ringmaster** with 0fb68ccf (400 playouts)
```
leelaz-next v leelaz-small-node (200/200 games)
board size: 19   komi: 7.5
                    wins              black          white       avg cpu
leelaz-next          101 50.50%       56  56.00%     45 45.00%     11.36
leelaz-small-node     99 49.50%       55  55.00%     44 44.00%     10.81
                                      111 55.50%     89 44.50%
```

**Ringmaster** with weights.txt (400 playouts)
```
leelaz-next v leelaz-small-node (200/200 games)
board size: 19   komi: 7.5
                    wins              black          white       avg cpu
leelaz-next          101 50.50%       55  55.00%     46 46.00%     13.50
leelaz-small-node     99 49.50%       54  54.00%     45 45.00%     12.81
                                      109 54.50%     91 45.50%
```

**Ringmaster** with weights.txt  (800 playouts on debug with -t4)
```
leelaz-next v leelaz-small-node (400/400 games)
board size: 19   komi: 7.5
                    wins              black          white        avg cpu
leelaz-next          202 50.50%       115 57.50%     87  43.50%     36.06
leelaz-small-node    198 49.50%       113 56.50%     85  42.50%     36.12
                                      228 57.00%     172 43.00%
```

Ringmaster had two failures earlier, I found a possible race condition with an assert and fixed that. I tested 600 more games with ringmaster and separately tested with gdb for 10 hours with no failures.

----
**Backup**

Edges (https://github.com/gcp/leela-zero/pull/878) theoretically requires 30% of original memory and in practice requires 60% of the old memory.

Unexpanded theoretically requires 15% of original memory and in practice requires 30% BUT comes at the cost of more complexity in `uct_select_child` and `UCTSearch` (having to call expand_child in a couple of extra places)

I believe the "practical limits" for playouts (on my UNIX 16 gigs machine) are
Next ~200K playouts
Edges ~300K playouts
Unexpanded ~1500K playouts

IMHO none of this really matters for self-play only for people who want to run large number of playouts.

|playouts|leelaz-next|leelaz-edges|leelaz-unexpanded|
|:---|---:|---:|---:|
|5,000|253144|199096|148932|
|50,000|1409680|874620|369920|
|250,000|6291676|3630720|1070976|

None of this has a notable impact on speed on either my fast GPU or slow GPU

----

**Peculiarities**
Currently `uct_select_child` depends on m_children sort order more than IMHO it should which makes verifying this is a no-op slightly harder. If you want to test

add to uct_select_child on line ~343 
```    if (parentvisits == 0) {
        // sort by score.
        parentvisits = 1;
    }
```

### CODE REVIEW

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -112,53 +111,55 @@ bool UCTNode::create_children(std::atomic<int> & nodecount,
...
+    m_child_scores.reserve(nodelist.size());
     for (const auto& node : nodelist) {
-        m_children.emplace_back(
-            std::make_unique<UCTNode>(node.second, node.first, init_eval)
-        );
+        m_child_scores.emplace_back(node.second, node.first);
     }
 
-    nodecount += m_children.size();
+    m_child_init_eval = init_eval;
```
Comment (WhiteHalmos) [inline src/UCTNode.cpp]: Rather than add m_child_init_eval, rebase on #888 to save this complexity.

File: src/UCTNode.cpp [hunk 2]
```diff
@@ -303,17 +308,43 @@ void UCTNode::accumulate_eval(float eval) {
     atomic_add(m_blackevals, (double)eval);
 }
 
-UCTNode* UCTNode::uct_select_child(int color) {
-    UCTNode* best = nullptr;
-    auto best_value = -1000.0f;
 
+// Expand nth child and remove from m_child_scores.
+UCTNode* UCTNode::expand(size_t child) {
```
Comment (WhiteHalmos) [inline src/UCTNode.cpp]: I think `expand` needs to be locked.

File: src/UCTNode.cpp [hunk 3]
```diff
@@ -322,13 +353,21 @@ UCTNode* UCTNode::uct_select_child(int color) {
...
 
-    for (const auto& child : m_children) {
-        if (!child->valid()) {
-            continue;
-        }
+    // positive index mean expanded nodes, negative index for unexpanded.
+    auto best_index = 0;
+    auto best_value = -1000.0f;
+
+    for (auto i = size_t{0}; i < m_expanded.size(); i++) {
```
Comment (WhiteHalmos) [inline src/UCTNode.cpp]: Rather than the complexity of positive and negative indices, just initialize best_index to -1 and use that to mean it's a child of m_expanded, and use `for (auto child : m_expanded)` instead. It's simpler than changing the meaning of index.

File: src/UCTNode.cpp [hunk 4]
```diff
@@ -376,43 +437,58 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
...
 UCTNode& UCTNode::get_best_root_child(int color) {
     LOCK(get_mutex(), lock);
-    assert(!m_children.empty());
 
-    return *(std::max_element(begin(m_children), end(m_children),
+    if (m_expanded.size() < 2) {
+        // NodeComp first checks visits so only need to consider m_expanded,
+        // unless few things has been expanded then for simplicity expand_all.
+        // This can be improved by expanding and returning node with max score.
+        expand_all();
```
Comment (WhiteHalmos) [inline src/UCTNode.cpp]: Not sure about the need to `expand_all` here. If m_expanded is empty, then this should just return the first child since all get_evals would be equal in that case.

**Comment by sethtroisi** [comment]
Looking at memory consumption of only one move isn't a good indication of how many playouts the game can be played with. Tree reuse means you carry on average ~40% of those playouts forward and even more when time management is off and responding to a sente move.

**leela-zero/next with --timemanagement off**

|playouts|benchmark|10 moves|full game (-r0)|
|:-----|----|----|----|
|1600|161|322|616
|5000|247|640|1693|
|10000|374|1124|3270|
|20000|624|2484||

------

**Games with high playouts**

timemanegement on, 250000 playouts
Maximum resident set size (kbytes): 2148272
http://eidogo.com/#1GoYKgms7

File: src/UCTNode.cpp (hunk 2 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: Done.

File: src/UCTNode.cpp (hunk 4 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: in fixing this I introduced a subtle bug which worries my, I'm backing out some of the clever optimizations and changing a number of places to call expand_all()

If we want to optimize performance later I think that's a fine thing but for now I'm more worried about making mistakes.

**Comment by gcp** [comment]
I'll pull #888 first, after that has been rebased.

**Comment by gcp** [comment]
Can you rebase this please?

**Comment by sethtroisi** [comment]
Rebased but please hold off on merging for now. I want to extensively test this PR as it has a number of edge cases.

File: src/UCTNode.cpp (hunk 1 above)
Comment (WhiteHalmos) [reply src/UCTNode.cpp]: @sethtroisi Friendly ping for removing m_child_init_eval now that m_net_eval is available.

File: src/UCTSearch.h [hunk 5]
```diff
@@ -73,12 +73,12 @@ class UCTSearch {
...
-        Maximum size of the tree in memory. Nodes are about
-        48 bytes, so limit to ~1.2G on 32-bits and about 5.5G
-        on 64-bits.
+        Maximum size of the tree in memory.
+        Nodes + Scores average less than 20 bytes.
+        limit to ~1.1G on 32-bits and about 5.6G on 64-bits.
     */
     static constexpr auto MAX_TREE_SIZE =
-        (sizeof(void*) == 4 ? 25'000'000 : 100'000'000);
+        (sizeof(void*) == 4 ? 60'000'000 : 300'000'000);
```
Comment (WhiteHalmos) [inline src/UCTSearch.h]: 60 million nodes for 32-bit, i.e. 4GB max RAM?

Comment (sethtroisi) [reply src/UCTSearch.h]: The original number was set to use ~ 1.2G of ram (see the now updated comment above).

**Comment by sethtroisi** [comment]
@WhiteHalmos I'm seeing modest performance gains (8-15%). Would you mind patching and comparing --benchmark (which I'm loving) with next branch?

File: src/UCTNode.cpp [hunk 6]
```diff
@@ -322,15 +362,25 @@ UCTNode* UCTNode::uct_select_child(int color) {
         }
     }
 
+    if (parentvisits == 0) {
+        // sort by score.
+        parentvisits = 1;
```
Comment (sethtroisi) [inline src/UCTNode.cpp]: I believe this is a no-op.

Before children had been sorted by score then would have equal eval and first child (with best score) would have been chosen.

File: src/UCTNode.cpp [hunk 7]
```diff
@@ -377,64 +441,96 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
...
+    assert(!m_expanded.empty());
+    std::stable_sort(rbegin(m_expanded), rend(m_expanded), NodeComp(color));
 }
 
 UCTNode& UCTNode::get_best_root_child(int color) {
     LOCK(get_mutex(), lock);
-    assert(!m_children.empty());
 
-    return *(std::max_element(begin(m_children), end(m_children),
+    // Check if any of m_expanded have visits.
```
Comment (Hersmunch) [inline src/UCTNode.cpp]: Can this loop be avoided by doing something like:
if expanded is empty return expanded best child
best_expanded = max(expanded)
if best_expanded visits == 0 return expanded best child
return best_expanded
?

File: src/UCTNode.cpp [hunk 8]
```diff
@@ -377,64 +441,96 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
...
+
 // Used to find new root in UCTSearch
 UCTNode::node_ptr_t UCTNode::find_child(const int move) {
+    LOCK(get_mutex(), lock);
+
     if (m_has_children) {
-        for (auto& child : m_children) {
+        for (auto& child : m_expanded) {
             if (child->get_move() == move) {
+                // This causes child to because null and subsequent calls to
```
Comment (Hersmunch) [inline src/UCTNode.cpp]: typo "because"?

File: src/UCTNode.cpp (hunk 7 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: Almost works,

when best_expanded.visits == 0 && best_expanded.score > best_unexpanded.score current code would return best_expanded, your pseudo code would return best_unexpanded

File: src/UCTSearch.cpp [hunk 9]
```diff
@@ -175,12 +175,19 @@ void UCTSearch::dump_stats(KoState & state, UCTNode & parent) {
...
-
-
-    if (parent.get_first_child()->first_visit()) {
+    auto first_child = parent.get_first_child();
+    if (first_child == nullptr || first_child->first_visit()) {
         return;
     }
 
+    if (parent.get_children().size() < 2) {
+        // There's a small change only one child has been expanded and debug
```
Comment (Hersmunch) [inline src/UCTSearch.cpp]: typo "change"

File: src/UCTNode.cpp [hunk 10]
```diff
@@ -304,16 +300,60 @@ void UCTNode::accumulate_eval(float eval) {
...
+
+    auto best = size_t{0};
+    auto bestScore = -1.0f;
+    for (auto i = size_t{0}; i < m_child_scores.size(); i++) {
+        auto score = m_child_scores[i].second;
+        if (score > bestScore) {
+            bestScore = score;
+            best = i;
+        }
+    };
```
Comment (Hersmunch) [inline src/UCTNode.cpp]: spurious ;

File: src/UCTNode.cpp (hunk 7 above)
Comment (Hersmunch) [reply src/UCTNode.cpp]: Whoops, so it does. In that case the best_expanded and best_unexpanded scores can be compared to decide which to return.

File: src/UCTNode.cpp [hunk 11]
```diff
@@ -322,15 +362,25 @@ UCTNode* UCTNode::uct_select_child(int color) {
...
 
-    for (const auto& child : m_children) {
-        if (!child->valid()) {
-            continue;
-        }
+    auto best_index = 0;
+    auto best_value = -1000.0f;
+
+    FastBoard board;
+    board.reset_board(19);
```
Comment (Hersmunch) [inline src/UCTNode.cpp]: What is this doing?

**Comment by WhiteHalmos** [comment]
Here's performance on my machine. About 6% increase in performance. I am slightly concerned that it doesn't generate the same results.

With node v2:
```
Thinking at most 2678400.0 seconds...
NN eval=0.507272

  D4 ->     563 (V: 51.55%) (N: 32.09%) PV: D4 Q4 D16 C3 D3 C4 C5 B5 B6 C6 D5 B7 B4 A6 B3 C17 D17 C16 C15 B15 B14 C14 D15 B13
 D16 ->     403 (V: 51.35%) (N: 25.21%) PV: D16 Q4 D4 C3 D3 C4 C5 B5 B6 C6 D5 B7 B4 A6 B3 C17 D17 C16 C15 B15 B14
  Q4 ->     265 (V: 51.36%) (N: 16.52%) PV: Q4 D4 D16 R3 R4 Q3 P3 P2 O2 O3 P4 N2 Q2 O1 R2 C17 D17 C16
 D17 ->      83 (V: 52.06%) (N:  3.30%) PV: D17 D4 Q4 R3 R4 Q3 P3 P2 O2
 R17 ->      68 (V: 51.05%) (N:  4.72%) PV: R17 R16 Q17 P17 P18 O18 O17 P16 N18 Q18 O19 R18 D16 Q4 D4
 C16 ->      63 (V: 51.85%) (N:  2.99%) PV: C16 D4 Q4 R3 Q3 R4 R5 S5 S6
  D3 ->      51 (V: 51.59%) (N:  2.66%) PV: D3 D16 Q4 R3 R4 Q3 P3 P2
  Q3 ->      41 (V: 51.68%) (N:  2.09%) PV: Q3 D4 D16 C17 C16 D17 E17 E18
  C4 ->      32 (V: 51.19%) (N:  2.15%) PV: C4 D16 Q4 R3 R4 Q3 P3 P2
  R4 ->      31 (V: 50.94%) (N:  2.23%) PV: R4 D4 D16 C17 C16 D17
1601 visits, 1961 nodes (563610 scores), 1600 playouts, 1103 n/s
```

With next
```
Thinking at most 2678400.0 seconds...
NN eval=0.507272

  D4 ->     563 (V: 51.55%) (N: 32.09%) PV: D4 Q4 D16 C3 D3 C4 C5 B5 B6 C6 D5 B7 B4 A6 B3 C17 D17 C16 C15 B15 B14 C14 D15 B13
 D16 ->     403 (V: 51.35%) (N: 25.21%) PV: D16 Q4 D4 C3 D3 C4 C5 B5 B6 C6 D5 B7 B4 A6 B3 C17 D17 C16 C15 B15 B14
  Q4 ->     265 (V: 51.36%) (N: 16.52%) PV: Q4 D4 D16 R3 R4 Q3 P3 P2 O2 O3 P4 N2 Q2 O1 R2 C17 D17 C16
 D17 ->      83 (V: 52.06%) (N:  3.30%) PV: D17 D4 Q4 R3 R4 Q3 P3 P2 O2
 R17 ->      68 (V: 51.05%) (N:  4.72%) PV: R17 R16 Q17 P17 P18 O18 O17 P16 N18 Q18 O19 R18 D16 Q4 D4
 C16 ->      63 (V: 51.85%) (N:  2.99%) PV: C16 D4 Q4 R3 Q3 R4 R5 S5 S6
  D3 ->      51 (V: 51.59%) (N:  2.66%) PV: D3 D16 Q4 R3 R4 Q3 P3 P2
  Q3 ->      41 (V: 51.68%) (N:  2.09%) PV: Q3 D4 D16 C17 C16 D17 E17 E18
  C4 ->      32 (V: 51.19%) (N:  2.15%) PV: C4 D16 Q4 R3 R4 Q3 P3 P2
  R4 ->      31 (V: 50.94%) (N:  2.23%) PV: R4 D4 D16 C17 C16 D17
1601 visits, 565571 nodes, 1600 playouts, 1038 n/s
```

**Comment by sethtroisi** [comment]
@WhiteHalmos I lost all of yesterday to tracking down mismatches in benchmark

I added this comment in the testing section:

Hand inspected the first `uct_select_child` instance that don't match with both gpu and cpu paths, difference is in 5th sig fig of winrate (AKA neural net eval) calculation

on such example
0.519917 = 0.503896 + 0.234812, 564 D16
0.519903 = 0.503882 + 0.234812, 564 D16

File: src/UCTNode.cpp (hunk 10 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: Fixed

File: src/UCTNode.cpp (hunk 11 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: I was using this for debug and it didn't get pulled out. Thanks for noticing

File: src/UCTSearch.cpp (hunk 9 above)
Comment (sethtroisi) [reply src/UCTSearch.cpp]: changed to chance

File: src/UCTSearch.h (hunk 5 above)
Comment (sethtroisi) [reply src/UCTSearch.h]: Those are **nodes**, max_tree_size counts both expanded and unexpanded nodes.
With 250,000 playouts, tree reuse, and timemanage off you'll probably see 400,000 playouts after some sente move.
~400,000 nodes * ~300 scores/node = 120,000,000 (nodes+scores) * 20 bytes = 2.4 GB


 there (360-move)

File: src/UCTNode.cpp (hunk 7 above)
Comment (sethtroisi) [reply src/UCTNode.cpp]: I think that would make the score sufficiently more complex (I'd either have to have a duplicate code of NodeComp logic or expand(best_child_score) either temporarily with logic to handle that or permanently in which case it's the same as this code.

File: src/UCTNode.cpp [hunk 12]
```diff
@@ -377,61 +439,97 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
...
-    return m_children.front().get();
+    if (!m_child_scores.empty()) {
+        return expand(best_child_score());
+    }
+    return nullptr;
 }
 
 const std::vector<UCTNode::node_ptr_t>& UCTNode::get_children() const {
-    return m_children;
+    return m_expanded;
```
Comment (sethtroisi) [inline src/UCTNode.cpp]: This method is no longer safe and fixing will take sometime

File: src/UCTNode.cpp (hunk 7 above)
Comment (Hersmunch) [reply src/UCTNode.cpp]: Agreed. Oh well.
Maybe I’ll have a play to see if there is any worthwhile performance gain by not expanding the best unexpanded to warrant e.g. making a function to compare an unexpanded to an expanded node that calls a function with the common code from NodeComp pulled out.
Shall I remove this review comment?

**Comment by gcp** [comment]
Going to close this a seth seems to be busy with other work, and some smart way of doing tree trimming may be more efficient / easier to implement.
